### ISSUE TITLE
2FA - Authenticator secret overridden upon start of 2FA setup

### ISSUE THREAD

**Issue by Kishi85** [jwag956/flask-security#238]
Due to the current non-atomic way tf_totp secret and tf_primary method are updated it is possible to override the tf_totp_secret with a new unvalidated value by doing the following:
1. Start 2FA setup with any method (sets tf_totp_secret)
2. Verify 2FA setup (sets tf_primary_method)
3. Re-start the 2FA setup with any method (resets tf_totp_secret)
4. Rather than finishing the setup by validating, decide that you don't want to scan the qrcode and just stop the process (for whatever reason).

At Step 4 the tf_totp_secret is already a different value than that validated by the user in Step 2 with tf_primary_method still being active and assumed valid. The user won't be able to login with the current token in his authenticator anymore.

Possible solutions:
1. Keep the tf_totp_secret and tf_primary_method changes atomic and at the end of the process (as already noted in #215). This is problematic because the session by default being serialized and stored in the session cookie, less so on custom session implementations storing data persistently on the server and just sending a session identifier in the cookie.
2. Always resetting the tf_primary_method when the tf_totp_secret is reset. Which is probably not that optimal either because the existing two-factor authentication will be reset when two-factor setup is started, regardless of result of the new setup.

The second solution won't work with mandatory 2FA, as tf_primary_method has to be reset to None whilst the user is validating the new tf_totp_secet. 
The first solution has the usual security problems with the new tf_totp_secret potentially leaking to the client with flasks default session mechanism (which could probably be mitigated by encrypting the whole string with something like flasks SECRET_KEY or something along those lines).

**Comment by jwag956** [comment]
Sigh - I actually had this happen to me while testing :-)

So - one thought is - do we really need to change the totp_secret? I believe I added that thinking it was a good idea (which has caused a lot of headaches). Could we simply set it once and not change it. If the account needs reseting - an admin can simply set both primary_method and totp_secret to None and the user can set up again. (might be nice to have a 'reset_tf' UserDatastore method to make the admin a bit easier.

**Comment by Kishi85** [comment]
I think the user should have a way to reset tf_totp_secret themselves if they realize it was compromised. the current way to do this is a good one as far as i'm concerned.

What about adding a temporary storage variable on the user object for the new tf_totp_secret lets call it 'new_tf_totp_secret' which then can be atomically committed to the real tf_totp_secret upon completion of the 2fa setup process? That way we could keep the reset behaviour while fixing this bug. The only new thing needed would be an additional variable/db colum on the user.

**Comment by Kishi85** [comment]
> I think the user should have a way to reset tf_totp_secret themselves if they realize it was compromised. the current way to do this is a good one as far as i'm concerned.
> 
> What about adding a temporary storage variable on the user object for the new tf_totp_secret lets call it 'new_tf_totp_secret' which then can be atomically committed to the real tf_totp_secret upon completion of the 2fa setup process? That way we could keep the reset behaviour while fixing this bug. The only new thing needed would be an additional variable/db colum on the user.

Probably something like this (just a quick untested hack):
```diff
diff --git a/flask_security/twofactor.py b/flask_security/twofactor.py
index 3df99e3..0f6afd2 100644
--- a/flask_security/twofactor.py
+++ b/flask_security/twofactor.py
@@ -151,6 +151,11 @@ def complete_two_factor_process(user, primary_method, is_changing, remember_logi
         user.tf_primary_method = primary_method
         _datastore.put(user)
 
+    if user.tf_totp_secret != user.new_tf_totp_secret:
+        user.tf_totp_secret = user.new_tf_totp_secret
+        user.new_tf_totp_secret = None
+        _datastore.put(user)
+
     # if we are changing two-factor method
     if is_changing:
         completion_message = "TWO_FACTOR_CHANGE_METHOD_SUCCESSFUL"
diff --git a/flask_security/views.py b/flask_security/views.py
index ed62afc..dcdd9cd 100644
--- a/flask_security/views.py
+++ b/flask_security/views.py
@@ -703,7 +703,7 @@ def two_factor_setup():
         # Regenerate the TOTP secret on every call of 2FA setup unless it is
         # within the same session and method (e.g. upon entering the phone number)
         if pm != session.get("tf_primary_method", None):
-            user.tf_totp_secret = generate_totp()
+            user.new_tf_totp_secret = generate_totp()
 
         session["tf_primary_method"] = pm
         session["tf_state"] = "validating_profile"
@@ -712,7 +712,7 @@ def two_factor_setup():
         _datastore.put(user)
         after_this_request(_commit)
 
-        send_security_token(user=user, method=pm, totp_secret=user.tf_totp_secret)
+        send_security_token(user=user, method=pm, totp_secret=user.new_tf_totp_secret)
         code_form = _security.two_factor_verify_code_form()
         if not _security._want_json(request):
             return _security.render_template(
@@ -961,7 +961,7 @@ def two_factor_qrcode():
     ):
         return abort(404)
 
-    totp = user.tf_totp_secret
+    totp = user.new_tf_totp_secret if user.new_tf_totp_secret else user.tf_totp_secret
     try:
         import pyqrcode
         identity_attributes = config_value("USER_IDENTITY_ATTRIBUTES")
```

**Comment by jwag956** [comment]
yup - I had thought about that as well - lets go with that then.

**Comment by jwag956** [comment]
Ok - I have been spending time re-reading passlib, and our implementation. Turns out that we (properly) encrypt the totp key as described:
https://passlib.readthedocs.io/en/stable/narr/totp-tutorial.html#encrypting-keys

which we did so that the key is encrypted in the DB. But that also means that possibly it isn't that bad to have it in the session cookie (temporarily) - since without the TWO_FACTOR_SECRET - it can't be decrypted.

That would make all this very simple.

**Comment by Kishi85** [comment]
Sounds good to me. Also i think that a note that recommends using a server-side session storage implementation (e.g. flask-session) for improved security if 2FA with tf_totp_secret is used should be added to the docs.

**Comment by Kishi85** [comment]
I've started working on this in https://github.com/Kishi85/flask-security/tree/tf_totp_secret-in-session

There are still 2 failing tests so far which I'm trying to figure out before starting a pull request. I've tried to leave the validation logic in forms alone, so those can be overridden/reimplemented easily with custom forms.
